[DOC] add a proposal for BlackboxExporter CRD (#7294)
prometheus-operator/prometheus-operator: Documentation/proposals/20250122-blackbox-crd.md

## Non-Goals

- This propose don't aim to replace the probe crd

## How

We propose creating two new Custom Resource Definitions (CRDs):
- The BlackboxExporter CRD, which functions similarly to the Prometheus CRD, will manage the deployment configuration and lifecycle of blackbox-exporter instances
- The BlackboxExporterModule CRD will define module specifications and automatically generate blackbox configurations, analogous to how Alertmanager works in Prometheus Operator

## Alternatives

- Update blackbox configuration whenever a probe module is modified or added

## Action Plan

1. Create the BlackboxExporter CRD which will manage blackbox-exporter instances as either a DaemonSet or Deployment
2. Implement a configuration reloader sidecar container to automatically reload blackbox-exporter when configuration changes, similar to how Prometheus Operator handles configuration updates

## CRDs Drafg

```yaml
apiVersion: monitoring.zhipu.ai/v1alpha1
kind: BlackboxExporter
metadata:
  name: blackbox-exporter-draft
spec:
  # Deployment mode - how the BlackboxExporter should be deployed 
    # Or Daemonset mode

>   deployMode: Deployment

Reviewer: is there a real need for a daemonset  mode? what would be the use case?

Reviewer 2: +1

  replicas: 2
  
  # Key configuration options

>   baseImage: prom/blackbox-exporter

Reviewer: baseImage is  deprecated, we should just have an `image` field.

  timeoutOffset: "0.5"
  logLevelProber: "info"
  
  # Module selection - allows selecting which probe modules to include

>   moduleSelector:

Reviewer: I assume that we want a moduleNamespaceSelector for consistency with other resources.

    matchLabels:
      type: http-probe
  

>   # Resource requirements

Reviewer: I'd be wary to include more deployment-specific fields to the CRD because from experience, we'll end up re-implementing the deployment spec. One option could be to let users customize the generated deployment as they want relying on [server-side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/#custom-resources-and-server-side-apply). I've used it in other projects and it works fine (provided that you run a recent Kubernetes version).

Reviewer: I was talking about all the fields which are specific to container definitions like `resources`

  resources:
    limits:
      memory: 200Mi
      cpu: 100m
    requests:
      memory: 100Mi
      cpu: 50m
      
  # Advanced configuration options
  securityContext:
    runAsNonRoot: true
    runAsUser: 1000
  
  # Network configuration
  webListenAddresses:
    - ":9115"
  externalURL: "https://blackbox.example.com"
  
status:
  # Status fields will be managed by the controller
  replicas: 2
  availableReplicas: 2
  updatedReplicas: 2
  unavailableReplicas: 0
  conditions:
    - type: Available
      status: "True"
      lastTransitionTime: "2025-02-28T12:00:00Z"
      reason: MinimumReplicasAvailable
      message: "All replicas are available"

---
apiVersion: monitoring.zhipu.ai/v1alpha1
kind: BlackboxExporterModule
metadata:
  name: example-http-probe
  labels:
    type: http-probe
spec:
  # HTTP probe configuration
  # Common probe settings
  timeout: 5s
  prober: http

>   http:

Reviewer 3: This resembles the BB configuration as is, but I was thinking maybe we can add something like, WDYT?

```yaml
prober: http // (http, tcp, dns, icmp, grpc)
proberConfg:
    http:
      method: GET
      headers:
        Host: host.example.com
    icmp:
      preferred_ip_protocol: "ip4"
```

Reviewer 3: Hi @Pger-Y you're right, I was thinking a bit about the current AlertmanagerConfig structure when I_ suggested this - https://doc.crds.dev/github.com/prometheus-operator/prometheus-operator/monitoring.coreos.com/AlertmanagerConfig/v1alpha1@v0.80.0. But we can just keep this simpler for a first version. But at the same time probably keeping as close as the current configuration could be a good idea, just in terms of structure, for context https://github.com/prometheus/blackbox_exporter/blob/master/example.yml

Reviewer: @Pger-Y the API structure should follow the general Kube API conventions. While it may seem counter-intuitive to not always follow the Prometheus config format, the Kube conventions exist for good reasons.

https://prometheus-operator.dev/docs/community/contributing/#changes-to-the-apis

* Teams wanting to manage blackbox module configurations declaratively through Kubernetes CRDs
* Users requiring a standardized, Kubernetes-native approach to monitor external targets

## Non-Goals

- This propose don't aim to replace the probe crd

## How

We propose creating two new Custom Resource Definitions (CRDs):
- The BlackboxExporter CRD, which functions similarly to the Prometheus CRD, will manage the deployment configuration and lifecycle of blackbox-exporter instances
- The BlackboxExporterModule CRD will define module specifications and automatically generate blackbox configurations, analogous to how Alertmanager works in Prometheus Operator

## Alternatives
Users may need to update or create a Probe CRD to monitor an HTTP endpoint with specific headers or body content. This often requires updating the Blackbox Exporter configuration, such as creating a new module or modifying an existing one. While users can manage this manually, it is a more cumbersome and error-prone process.

## Action Plan

1. Create the BlackboxExporter CRD which will manage blackbox-exporter instances as either a DaemonSet or Deployment

> 2. Implement a configuration reloader sidecar container to automatically reload blackbox-exporter when configuration changes, similar to how Prometheus Operator handles configuration updates

Reviewer: is there a need for another config reloader? I imagine that we can reuse the existing one.

> ## CRDs Draft

Reviewer: (nit) this should go in the `How` section

### 1. BlackBoxExporter CRD
```yaml

> apiVersion: monitoring.zhipu.ai/v1alpha1

Reviewer: (nit) wrong group

kind: BlackboxExporter
metadata:
  name: blackbox-exporter-draft
spec:
  # Deployment mode - how the BlackboxExporter should be deployed 
    # Or Daemonset mode
  deployMode: Deployment
  replicas: 2
  
  # Key configuration options
  image: prom/blackbox-exporter
  timeoutOffset: "0.5"

>   logLevelProber: "info"

Reviewer: we need fields for `--log.level`, `--log.format` and `--history.limit` too.

  
  # Module selection - allows selecting which probe modules to include
  moduleNamespaceSelector:
    matchLabels:
        kubernetes.io/metadata.name: monitoring
  moduleSelector:
    matchLabels:
      type: http-probe
  
  # Resource requirements
  resources:
    limits:
      memory: 200Mi
      cpu: 100m
    requests:
      memory: 100Mi
      cpu: 50m
      
  # Advanced configuration options
  securityContext:
    runAsNonRoot: true
    runAsUser: 1000
  
  # Network configuration

>   webListenAddresses:

Reviewer: I'd recommend defining all web fields in the same group

````
spec:
  web:
    listenAddresses: [...]
    externalURL: ...
    ...
```

    - ":9115"
  externalURL: "https://blackbox.example.com"
  
status:
  # Status fields will be managed by the controller
  replicas: 2
  availableReplicas: 2
  updatedReplicas: 2
  unavailableReplicas: 0
  conditions:
    - type: Available
      status: "True"
      lastTransitionTime: "2025-02-28T12:00:00Z"
      reason: MinimumReplicasAvailable
      message: "All replicas are available"

```
### 2. BlackBoxExporterModule CRD
```
---

> apiVersion: monitoring.zhipu.ai/v1alpha1

Reviewer: same here

kind: BlackboxExporterModule
metadata:
  name: example-http-probe
  labels:
    type: http-probe
spec:
  # HTTP probe configuration
  # Common probe settings
  timeout: 5s
  prober: http
  http:
    # Request definition
    method: POST
    headers:

>       Accept: "application/json"

Reviewer: see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps
maps aren't recommended in Kubernetes APIs.

## Add BlackboxExporter and BlackboxExporterModule CRDs

* **Owners:**
  * [Pger-Y](https://github.com/Pger-Y)

* **Related Tickets:**
  * [Discussion](https://github.com/prometheus-operator/prometheus-operator/discussions/7134#discussioncomment-11569949)

* **Other docs:**
  * n/a

This document proposes the creation of a Blackbox Operator and associated CRDs, similar to Prometheus-Operator, to manage blackbox-exporter instances and introduce a Module CRD for handling dynamic configuration of blackbox modules in a Kubernetes-native way.

## Why

> As the above discussion described the Probe CRD is extremely convenient for icmp prober, the Probe CRD can define anything icmp needed ,but when we want to use http Probe , we must define header/body and plenty of something else in blackbox config first,because http is much more complex than icmp.

Reviewer 3: Hi @Pger-Y just a nit pic about punctuation:

```MD
As the above discussion described the Probe CRD is extremely convenient for icmp prober, the Probe CRD can define anything icmp needed, but when we want to use http Probe we must define header/body and plenty of something else in blackbox config first, because http is much more complex than icmp.

The situation is very similar to ScrapeConfig/ServiceMonitor that prometheus-operator does. We need a k8s way for generate blackbox config dynamic by reading the Probe Module Config 
```

> The situation is very similar to ScrapeConfig/ServiceMonitor that prometheus-operator does,We need a k8s way for generate blackbox config dynamic by reading the Probe Module Config 

Reviewer 3: Also I do believe the ScrapeCRD|ServiceMonitor is perhaps not the best comparison, since by design was done to add [additional service discover capabilities to targets outside k8s](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/proposals/202212-scrape-config.md#why). IMO Alertmanager|AlertmanagerConfig is a better comparison

### Pitfalls of the current solution

The Probe CRD provides a partial solution for handling complex modules like HTTP and gRPC. However, it introduces a separate lifecycle for Blackbox management. Users must pre-define modules containing specific configurations (such as headers and request bodies) by manually updating and reloading the Blackbox configuration, before referencing these modules in their probes.

## Goals

- Provide a way for users to self-service adding module config for their Probe CRD
- Consolidate the module configuration generation logic in a central point for other resources to use

### Audience

* Administrators providing Prometheus monitoring who want to enable self-service probe configuration for their users
* Teams wanting to manage blackbox module configurations declaratively through Kubernetes CRDs
* Users requiring a standardized, Kubernetes-native approach to monitor external targets

## Non-Goals

- This propose don't aim to replace the probe crd

## How

> As discussed in [Discussion](https://github.com/prometheus-operator/prometheus-operator/discussions/7134#discussioncomment-11569949), we propose creating two new Custom Resource Definitions (CRDs):

Reviewer: (nit) I'd remove the link here.
```suggestion
We propose creating two new Custom Resource Definitions (CRDs):
```

- The BlackboxExporter CRD, which functions similarly to the Prometheus CRD, will manage the deployment configuration and lifecycle of blackbox-exporter instances

> - The BlackboxExporterModule CRD will define module specifications and automatically generate blackbox configurations, analogous to how ScrapeConfig works in Prometheus Operator

Reviewer 3: I guess it needs to be changed here as well

## Alternatives

- Update blackbox configuration whenever a probe module is modified or added

## Action Plan

1. Create the BlackboxExporter CRD which will manage blackbox-exporter instances as either a DaemonSet or Deployment
2. Implement a configuration reloader sidecar container to automatically reload blackbox-exporter when configuration changes, similar to how Prometheus Operator handles configuration updates

> ## CRDs draft

Reviewer: Thanks for adding the CRD structure. Having said that it's too detailed for a design proposal. In general, it's sufficient to add an example of the custom resource with the most important fields.

## Your Proposal Title

* **Owners:**
  * [Pger-Y](https://github.com/Pger-Y)

* **Related Tickets:**
  * [Discussion](https://github.com/prometheus-operator/prometheus-operator/discussions/7134#discussioncomment-11569949)

* **Other docs:**
  * n/a

> This document proposes the creation of a Blackbox Operator and associated CRDs, similar to Prometheus-Operator, to manage blackbox-exporter instances and introduce a Module CRD for handling dynamic configuration of blackbox modules in a Kubernetes-native way.

Reviewer 4: I don't see a reason for this to be a new operator. prometheus-operator already manages AlertManager and ThanosRuler resources for example.

Reviewer 5: I think he means operator in sense of adding this capability to the actual prometheus operator, this might be confuse especially looking for our structure, where you have operator.go inside many different resources like Prometheus, Thanos and many others.

> This document proposes the creation of a Blackbox Operator and associated CRDs, similar to Prometheus-Operator, to manage blackbox-exporter instances and introduce a Module CRD for handling dynamic configuration of blackbox modules in a Kubernetes-native way.

Reviewer 3: Hi @Pger-Y question, when you have "Kubernetes-native way", do you mean to be able to pick targets with selectors, or is it another thing I'm missing? Just because I've worked with the probe in the past and the downside at the time was the [static](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api-reference/api.md#monitoring.coreos.com/v1.ProbeTargetStaticConfig) config for targets.

Reviewer 3: Ok thanks I was just considering the targets since for my specific use case there were targets inside and outside K8s, and maybe expanding on the current Probe CRD would be a good idea. But I'm guessing that should be better suited for an issue instead.

## Why
As the above discussion described the Probe CRD is extremely convenient for icmp prober, the Probe CRD can define anything icmp needed ,but when we want to use http Probe , we must define header/body and plenty of something else in blackbox config first,because http is much more complex than icmp.

The situation is very similar to ScrapeConfig/ServiceMonitor that prometheus-operator does,We need a k8s way for generate blackbox config dynamic by reading the Probe Module Config 

### Pitfalls of the current solution

> Using HTTP Probe with custom header and body,we must modify the confgiMap of config file and reload blackbox by hand which is not that much cloud native :(,this situation is really a copy of prometheus-operator does

Reviewer 5: I'd extend to something like

The ProbeCRD partially solves the current solution when users require the utilisation of complex modules, such as HTTP, GRPC and many others, requiring users have a different life cycle of the Blackbox management`

Just a general idea.

## Goals

> - Provide a way for users to self-service adding module config for their Probe CRD

Reviewer 6: I feel it make more sense to have a BlackboxExporter CRD and BlackboxExporterModule CRD similar to Prometheus and PrometheusRule or Alertmanager and AlertmanagerConfig

- Consolidate the module configuration generation logic in a central point for other resources to use

### Audience

* Administrators providing Prometheus monitoring who want to enable self-service probe configuration for their users
* Teams wanting to manage blackbox module configurations declaratively through Kubernetes CRDs
* Users requiring a standardized, Kubernetes-native approach to monitor external targets

## Non-Goals

> * This proposal aims to be an expansion of the Probe CRD rather than replacing it.

Reviewer 5: I feel this is not a right definition of a Non-Goal.

Maybe we can rewrite to be: This propose don't aim to replace the probe crd.

## How

> As discussed in [Discussion](https://github.com/prometheus-operator/prometheus-operator/discussions/7134#discussioncomment-11569949), we propose to create a new Blackbox CRD that will function similarly to the Prometheus CRD. This controller will be responsible for managing blackbox-exporter instances and handling dynamic configuration updates and reloading.

Reviewer 5: I tend to agree with this separation of concerns especially because you might hand-over the configuration of modules to other teams, like we do with Monitor Objects.

The only down (or not) side is the fact of BlackBoxCRD will be minimal

## Alternatives

- Update blackbox configuration whenever a probe module is modified or added

> ## Action Plan

Reviewer 5: Thanks for starting this @Pger-Y 
I'd love to see some drafts of the CRDs to discuss an initial API Definition

## Why

> In our daily use, we have discovered that the Probe CRD is extremely convenient for ICMP prober. The Probe CRD can define anything ICMP needed, but when we want to use HTTP Probe, we must define header/body and plenty of other parameters in blackbox config first, because HTTP is much more complex than ICMP.

Reviewer 2: (nit)
```suggestion
The Probe CRD can define all the parameters that the ICMP Probe requires and is quite convenient for most use cases. But configuring the HTTP Probe using the Probe CRD is a lot more complex, requiring the user to define plenty of parameters. This makes the Probe CRD an insufficient solution and potentially error-prone for HTTP Probe use cases.
```

> The situation is very similar to Alertmanager/AlertmanagerConfig that prometheus-operator does. We need a Kubernetes-native way to generate blackbox config dynamically by reading the Probe Module Config.

Reviewer 2: (nit)
```suggestion
The situation is very similar to Alertmanager/AlertmanagerConfig that prometheus-operator offers. We need a Kubernetes-native way to generate blackbox config dynamically by reading the Probe Module Config.
```

## How

We propose creating two new Custom Resource Definitions (CRDs):
- The BlackboxExporter CRD, which functions similarly to the Prometheus CRD, will manage the deployment configuration and lifecycle of blackbox-exporter instances

> - The BlackboxExporterModule CRD will define module specifications and automatically generate blackbox configurations, analogous to how Alertmanager works in Prometheus Operator

Reviewer 2: Might be a stupid question. But isn't generating a configuration based on the CRD more relevant to the `ScrapeConfig` - `Prometheus` relationship?

Reviewer 3: Well IMO considering how BB works and it needs the prober with the prober type and relevant configuration for each probe I think is more a relation if configuration, so I think is more closer related to Alertmanager|AlermanagerConfig

1. Create the BlackboxExporter CRD which will manage blackbox-exporter instances as either a DaemonSet or Deployment
2. Implement a configuration reloader sidecar container to automatically reload blackbox-exporter when configuration changes, similar to how Prometheus Operator handles configuration updates

> ## CRDs Drafg

Reviewer 2: Typo :)
```suggestion
## CRDs Draft
```

> 

Reviewer 2: (suggestion)
### 1. `BlackBoxExporter` CRD

                description: |-
                  INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
                  Important: Run "make" to regenerate code after modifying this file
                format: int32
                type: integer
              selector:
                description: The selector used to match the pods targeted by this
                  Alertmanager object.
                type: string
              unavailableReplicas:
                description: Total number of unavailable pods targeted by this Alertmanager
                  object.
                format: int32
                type: integer
              updatedReplicas:
                description: |-
                  Total number of non-terminated pods targeted by this Alertmanager
                  object that have the desired version spec.
                format: int32
                type: integer
            required:
            - availableReplicas
            - replicas
            - unavailableReplicas
            - updatedReplicas
            type: object
        type: object
    served: true
    storage: true
    subresources:
      status: {}
```

> ## 2. `Module` CRD

Reviewer 2: keeping this seems reasonable.
### 2. `BlackBoxExporterModule` CRD
